### PR TITLE
[vcpkg] Honor CMAKE_DISABLE_FIND_PACKAGE_*. Fixes #18631

### LINKED ISSUE #18631: `find_package` in `vcpkg.cmake` must honor `CMAKE_DISABLE_FIND_PACKAGE_<NAME>`
**Describe the bug**

[In pristine CMake](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html), "every non-`REQUIRED` `find_package()` call in a project can be disabled by setting the variable `CMAKE_DISABLE_FIND_PACKAGE_<PackageName>` to `TRUE`." This means, there are no side effects.

`vcpkg.cmake` overrides cmake's original `find_package`. It always includes `vcpkg-cmake-wrapper.cmake` for `<PackageName>` if it exists:
https://github.com/microsoft/vcpkg/blob/876e67c26e42c0c7d2daa41c6871af117ae6bec5/scripts/buildsystems/vcpkg.cmake#L716-L725
So all side effects from the wrapper always occur, e.g. failures to load transitive dependencies.

**Actual behaviour:**
An example of the actual behaviour is seen in current simage build errors for osx: The simage build system is configured with `-DCMAKE_DISABLE_FIND_PACKAGE_TIFF=ON`. Still, *iff port tiff happens to be already installed* (and quicktime is not available), the wrapper for TIFF is loaded, and it causes an configuration error because the dependencies of TIFF (JPEG, ZLIB) are disabled, too:

```
-- Configuring x64-osx-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:105 (message):
    Command failed: /Users/vagrant/Data/downloads/tools/cmake-3.20.2-osx/cmake-3.20.2-macos-universal/CMake.app/Contents/bin/cmake /Users/vagrant/Data/buildtrees/simage/src/mage-1.8.0-17f988a113.clean -DSIMAGE_BUILD_SHARED_LIBS=OFF -DSIMAGE_USE_AVIENC=OFF -DSIMAGE_USE_GDIPLUS=OFF -DCMAKE_DISABLE_FIND_PACKAGE_FLAC=ON -DCMAKE_DISABLE_FIND_PACKAGE_Jasper=ON -DCMAKE_DISABLE_FIND_PACKAGE_Doxygen=ON -DCMAKE_DISABLE_FIND_PACKAGE_ZLIB=ON -DCMAKE_DISABLE_FIND_PACKAGE_GIF=ON -DCMAKE_DISABLE_FIND_PACKAGE_JPEG=ON -DCMAKE_DISABLE_FIND_PACKAGE_PNG=ON -DCMAKE_DISABLE_FIND_PACKAGE_TIFF=ON -DCMAKE_MAKE_PROGRAM=/Users/vagrant/Data/downloads/tools/ninja-1.10.2-osx/ninja -DCMAKE_SYSTEM_NAME=Darwin -DBUILD_SHARED_LIBS=OFF -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=/Users/vagrant/Data/work/1/s/scripts/toolchains/osx.cmake -DVCPKG_TARGET_TRIPLET=x64-osx -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=external -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=/Users/vagrant/Data/work/1/s/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=/Users/vagrant/Data/work/1/s -D_VCPKG_INSTALLED_DIR=/Users/vagrant/Data/installed -DVCPKG_MANIFEST_INSTALL=OFF -DCMAKE_OSX_ARCHITECTURES=x86_64 -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/Users/vagrant/Data/packages/simage_x64-osx/debug
```

```
CMake Error at /Users/vagrant/Data/installed/x64-osx/share/jpeg/vcpkg-cmake-wrapper.cmake:1 (_find_package):
  _find_package for module JPEG called with REQUIRED, but
  CMAKE_DISABLE_FIND_PACKAGE_JPEG is enabled.  A REQUIRED package cannot be
  disabled.
Call Stack (most recent call first):
  /Users/vagrant/Data/work/1/s/scripts/buildsystems/vcpkg.cmake:745 (include)
  /Users/vagrant/Data/installed/x64-osx/share/tiff/vcpkg-cmake-wrapper.cmake:17 (find_package)
  /Users/vagrant/Data/work/1/s/scripts/buildsystems/vcpkg.cmake:745 (include)
  CMakeLists.txt:222 (find_package)


CMake Error at /Users/vagrant/Data/work/1/s/scripts/buildsystems/vcpkg.cmake:791 (_find_package):
  _find_package for module ZLIB called with REQUIRED, but
  CMAKE_DISABLE_FIND_PACKAGE_ZLIB is enabled.  A REQUIRED package cannot be
  disabled.
```
(From  https://github.com/microsoft/vcpkg/issues/18393#issuecomment-861193370)

**Expected behavior:**
For consistency with CMake, no side effects are allowed for non-`REQUIRED` `find_package()` calls when the variable `CMAKE_DISABLE_FIND_PACKAGE_<PackageName>` is set to `TRUE`, i.e. `find_package` must return early, without loading wrappers.

In the simage example, the TIFF wrapper must not be loaded for the `find_package(TIFF)` call in simage's CMakeLists.txt, due to 
`CMAKE_DISABLE_FIND_PACKAGE_TIFF` being set.

### PR DESCRIPTION
**Pull request by ras0219-msft**
Fixes #18631. See that issue for full details.

### CODE REVIEW

**Comment by dg0yt** [comment]
LGTM. But as usual with `vcpkg.cmake`, this is a no-op in CI. Perhaps you want to increment the port-version of some relevant port (zlib, tiff)? Or move it to some of the audit rollups or CI re-provisioning?

**Review by BillyONeal — DISMISSED** [review]

**Review by JackBoosY — DISMISSED** [review]

File: scripts/ports.cmake [hunk 1]
```diff
@@ -1,5 +1,5 @@
-# rebuild: 2
-cmake_minimum_required(VERSION 3.20)
+# rebuild: 1
+cmake_minimum_required(VERSION 3.21)
```
Comment (dg0yt) [inline scripts/ports.cmake:2]: There is still 3.20.4 for freebsd in 
https://github.com/microsoft/vcpkg/blob/33f02c0ae50c262da487d21ace4f5d67ae949c18/scripts/vcpkgTools.xml#L31-L37

Comment (dg0yt) [reply scripts/ports.cmake:2]: CMake 3.21.4 seems to be available for freebsd now:
https://www.freebsd.org/cgi/ports.cgi?query=cmake&stype=name&sektion=all

Comment (dg0yt) [reply scripts/ports.cmake:2]: My comment is no longer relevant.
freebsd is still at 3.20.3 in vcpkg, but the requirement was bumped with another PR, and nobody (except for me) tool notice.

**Comment by dg0yt** [comment]
What is blocking this PR? I just stumbled over this again.

**Comment by BillyONeal** [comment]
> What is blocking this PR? I just stumbled over this again.

Robert (@ras0219) has a tendency to get dragged in to other things; he has not indicated recently that he is working on this. I imagine he would love to have this merged but he does not have the capacity right now to deal with the downstream fallout.

**Comment by dg0yt** [comment]
If resolving the merge conflicts properly, it should also fix #28552.

**Comment by dan-shaw** [comment]
Closing this PR, since the last update was over a year ago, please open a new PR if you are still working on this. Thanks!

**Comment by dg0yt** [comment]
Oh!

**Pull request closed without merging**